asm/kanread.s

```diff
@@ -0,0 +1,209 @@
+#include "ultra64/asm.h"
+
+// assembler directives
+.set noat
+.set noreorder
+
+.section .text
+
+.balign 16
+
+/**
+ * s32 Kanji_OffsetFromShiftJIS(s32 shiftJISCodepoint);
+ *
+ * Returns the offset of the glyph texture data in the file `kanji` corresponding
+ * to a given 2-byte Shift-JIS codepoint. No range validity check is carried out.
+ *
+ * A nice Shift-JIS codepoint table: https://uic.io/en/charset/show/shift_jis/
+ * The file `kanji` contains the 'Level 1' kanji (0x889F-0x9872), and a reworked
+ * version of the non-kanji section that includes extra English and Hylian glyphs.
+ *
+ * @param shiftJISCodepoint Codepoint of glyph.
+ * return s32 offset into `kanji` file.
+ *
+ * @remark Original name: "LeoGetKadr"
+ */
+LEAF(Kanji_OffsetFromShiftJIS)
+    // Characters with codepoints >= 0x8800 are kanji. Arrangement is regular, 
+    // so convert index directly.
+    li      $at, 0x8800
+    slt     $at, $a0, $at
+    bnez    $at, .nonkanji
+     // 0xBC is number of glyphs in one block in the `kanji` file:
+     // 0x100 possible codepoints with the same byte1
+     // - 0x40 unused at beginning 
+     // - 1 unused at 0x7F 
+     // - 3 unused at 0xFD, 0xFE, 0xFF
+     li     $a2, 0xBC
+    // Get byte1 and adjust so starts at 0
+    srl     $a1, $a0, 8
+    addi    $a1, $a1, -0x88
+    multu   $a2, $a1
+    // Get byte2 and adjust so starts at 0
+    andi    $a3, $a0, 0xff
+    addi    $a3, $a3, -0x40
+    slti    $at, $a3, 0x40
```

Comment:
This could be
```suggestion
    slti    $at, $a3, (0x7F + 1 - 0x40)
```
to highlight that it's checking against 0x7F but after the `- 0x40`
but idk, no big deal

Reply:
I feel like `stli ... (0x80 - 0x40)` might be a bit clearer since it's a < rather than <=, but I suppose it's constrained by the available instructions. Of course it never checks for 0x7F itself anyway.